benchmarks/replay/replay.cpp

```diff
@@ -74,42 +79,128 @@ struct allocation {
    *
+   * @param factory A factory function to create the memory resource
    * @param events The set of allocation events to replay
    * @param args Variable number of arguments forward to the constructor of MR
    */
   replay_benchmark(MRFactoryFunc factory,
                    std::vector<std::vector<rmm::detail::event>> const& events)
-    : mr_{factory()}, events_{events}
+    : factory_{factory}, mr_{}, events_{events}, allocation_map{events.size()}, event_index{0}
+  {
+  }
+
+  /**
+   * @brief Copy construct a replay_benchmark
+   *
+   * Does not copy the mutex or the map
+   */
+  replay_benchmark(replay_benchmark const& other)
+    : factory_{other.factory_},
+      mr_{other.mr_},
+      events_{other.events_},
+      allocation_map{events_.size()},
+      event_index{0}
+  {
+  }
+
+  /// Add an allocation to the map (thread safe)
+  void set_allocation(std::unordered_map<uintptr_t, allocation>& allocation_map,
+                      uintptr_t ptr,
+                      allocation alloc)
+  {
+    std::lock_guard<std::mutex> lock(map_mutex);
+    allocation_map.insert({ptr, alloc});
+  }
+
+  /// Remove an allocation from the map (thread safe)
+  allocation remove_allocation(std::unordered_map<uintptr_t, allocation>& allocation_map,
+                               uintptr_t ptr)
   {
+    std::lock_guard<std::mutex> lock(map_mutex);
+    auto iter = allocation_map.find(ptr);
+    if (iter != allocation_map.end()) {
+      allocation a = iter->second;
+      allocation_map.erase(iter);
+      return a;
+    }
+    return allocation{};
+  }
+
+  /// Create the memory resource shared by all threads before the benchmark runs
+  void SetUp(const ::benchmark::State& state) override
+  {
+    if (state.thread_index == 0) {
+      rmm::logger().log(spdlog::level::info, "------ Start of Benchmark -----");
+      mr_ = factory_();
+    }
   }
 
-  void operator()(benchmark::State& state)
+  /// Destroy the memory resource and count any unallocated memory
+  void TearDown(const ::benchmark::State& state) override
   {
-    // Maps a pointer from the event log to an active allocation
-    std::unordered_map<uintptr_t, allocation> allocation_map(events_.size());
+    if (state.thread_index == 0) {
+      rmm::logger().log(spdlog::level::info, "------ End of Benchmark -----");
+      // clean up any leaked allocations
+      std::size_t total_leaked{0};
+      std::size_t num_leaked{0};
+      for (auto const& ptr_alloc : allocation_map) {
+        auto alloc = ptr_alloc.second;
+        num_leaked++;
+        total_leaked += alloc.size;
+        mr_->deallocate(alloc.p, alloc.size, 0);
+      }
+      if (num_leaked > 0)
+        std::cout << "LOG shows leak of " << num_leaked << " allocations of " << total_leaked
+                  << " total bytes\n";
+      allocation_map.clear();
+      mr_.reset();
+    }
+  }
 
+  /// Run the replay benchmark
+  void BenchmarkCase(::benchmark::State& state) override
```

Comment:
I can't find `BenchmarkCase` documented anywhere and am not sure what it is for as opposed to just using the `operator()` directly. AFAIK, you can pass any callable object to `RegisterBenchmark`.

Reply:
`BenchmarkFixture::Run` is basically `SetUp(); BenchmarkCase(state); TearDown();`.   My operator() calls Run. I can change it so that the operator just calls setup and teardown I guess.  In that case it doesn't even need to be a fixture...

Reply:
Fixed.